File: Solution438/Solution438.java
package Solution438;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
    438. Find All Anagrams in a String
    Given two strings s and p, return an array of all the start indices of p's anagrams in s. You may return the answer in any order.
*/
class Solution438 {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        if (s.length() < p.length()) return result;

        int[] pCount = new int[26];
        int[] sCount = new int[26];

        for (int i = 0; i < p.length(); i++) {
            pCount[p.charAt(i) - 'a']++; // ASCII 
            sCount[s.charAt(i) - 'a']++; 
        }

        if(Arrays.equals(pCount, sCount)) result.add(0); //first window
        
        for (int i = p.length(); i<s.length(); i++) {
            sCount[s.charAt(i) - 'a']++; // add
            sCount[s.charAt(i - p.length()) - 'a']--; // remove

            if (Arrays.equals(pCount, sCount)) {
                result.add(i - p.length() + 1);
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Solution438 sol = new Solution438();
        System.out.println(sol.findAnagrams("abab", "ab"));
    }
}

File: Solution438/Solution438.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution438 {
public:
    vector<int> findAnagrams(string s, string p) {
        int s_len = s.length();
        int p_len = p.length();

        if (s_len < p_len) return {};
        
        vector<int> result;
        vector<int> pCount(26, 0);
        vector<int> sCount(26, 0);
        
        for (int i = 0; i < p_len; i++) {
            pCount[p[i] - 'a']++;
            sCount[s[i] - 'a']++;
        }
        if (pCount == sCount) result.push_back(0);

        for (int i = p_len; i < s_len; i++) {
            sCount[s[i] - 'a']++;
            sCount[s[i - p_len] - 'a']--;

            if (pCount == sCount) {
                result.push_back(i - p_len + 1);
            }
        }
        return result;
    }
};

int main() {
    Solution438 sol;
    string s = "abab";
    string p = "ab";
    cout << "input: s=" << s << ",p=" << p << endl;
    vector<int> res = sol.findAnagrams(s, p);
    cout << "result: ["; 
    for (int x : res) {
        cout << x << " ";
    }
    cout << "]" << endl;
    return 0;
}